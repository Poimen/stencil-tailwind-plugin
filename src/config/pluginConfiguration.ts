import path from 'path';
import fs from 'fs-extra';
import { Config as TailwindConfig } from 'tailwindcss';
import { PluginConfigOpts, PluginConfigOptsDefaults, TailwindPluginConfig, TailwindPluginFunctionalConfig } from '../index';
import { warn } from '../debug/logger';

function makeDefaultTailwindConf(): TailwindConfig {
  return {
    content: [],
    theme: {
      extend: {}
    },
    plugins: []
  };
}

function buildTailwindConfigurationBasedOnUserObject(userConfiguration: TailwindConfig): TailwindPluginFunctionalConfig {
  return (_: string, config: TailwindConfig): TailwindConfig => {
    // eslint-disable-next-line dot-notation
    const includePreflight = config.corePlugins['preflight'];
    let preflight = includePreflight;

    if (Array.isArray(userConfiguration.corePlugins)) {
      // user configuration is disabling all core plugins...
      return {
        ...userConfiguration,
        ...config,
        corePlugins: []
      };
    }

    if (includePreflight && userConfiguration.corePlugins) {
      // if we are including the preflight, resolve to the user configuration value if available
      // eslint-disable-next-line dot-notation
      preflight = userConfiguration.corePlugins['preflight'] ?? true;
    }
    return {
      ...userConfiguration,
      content: config.content,
      corePlugins: {
        ...userConfiguration.corePlugins,
        preflight
      }
    };
  };
}

export function resolveTailwindConfigurationFromUserSettings(tailwindConf: TailwindPluginConfig, contentFileList: string[], includePreflight: boolean): TailwindConfig {
  // Resolve the Tailwind configuration based on the user function
  const tailwindResolver = tailwindConf as TailwindPluginFunctionalConfig;
  return tailwindResolver(contentFileList[0], {
    content: contentFileList,
    corePlugins: {
      preflight: includePreflight
    }
  });
}

export const PluginConfigDefaults: PluginConfigOptsDefaults = {
  DEFAULT: {
    enableDebug: false,
    tailwindCssPath: undefined,
    tailwindCssContents: '@tailwind base;@tailwind utilities;@tailwind components;',
    tailwindConf: makeDefaultTailwindConf(),
    stripComments: false,
    minify: true,
    useAutoPrefixer: true
  }
};

function fetchTailwindCssContents(tailwindCssPath?: string): string | null {
  const tailwindCssContents = fs.readFileSync(path.resolve(tailwindCssPath)).toString();

  if (tailwindCssContents.length === 0) {
    warn('No css found when reading Tailwind configuration css - path', tailwindCssPath);
    return null;
  }

  return tailwindCssContents;
}

export function configurePluginOptions(opts: PluginConfigOpts): PluginConfigOpts {
  const configuration = Object.assign({}, opts);

  if (configuration.tailwindCssPath) {
    configuration.tailwindCssContents = fetchTailwindCssContents(configuration.tailwindCssPath) ?? configuration.tailwindCssContents;
  }

  if (typeof configuration.tailwindConf !== 'function') {
    const userConfiguration = configuration.tailwindConf as TailwindConfig;
    configuration.tailwindConf = buildTailwindConfigurationBasedOnUserObject(userConfiguration);
  }

  return configuration;
}
