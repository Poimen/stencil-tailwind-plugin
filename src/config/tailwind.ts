import { TailwindConfig } from 'tailwindcss/tailwind-config';

export function makeTailwindConfig(purgeFileList: string[]): TailwindConfig {
  const twConf = {
    mode: 'jit',
    purge: purgeFileList,
    darkMode: false,
    plugins: []
  };

  // HACK! Current tailwind types don't have the 'jit' mode field so make the type
  // look as if it is meant to be there
  return (twConf as unknown) as TailwindConfig;
}
