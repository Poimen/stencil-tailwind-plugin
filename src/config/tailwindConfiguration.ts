import path from 'path';
import fs from 'fs-extra';
import { TailwindConfig } from 'tailwindcss/tailwind-config';
import * as log from '../debug/logger';

let _tailwindCssContents: string;
let _tailwindConf: TailwindConfig;

export function makeTailwindConfig(purgeFileList: string[]): TailwindConfig {
  _tailwindConf.purge = purgeFileList;
  return _tailwindConf;
}

export function fetchTailwindDefaultCssConf(): string {
  return _tailwindCssContents;
}

function setDefaultTailwindCssContents(tailwindCssPath?: string) {
  const _defaultPackageTailwindCss = `
    @tailwind utilities;
    @tailwind components;
  `;

  if (tailwindCssPath) {
    _tailwindCssContents = fs.readFileSync(path.resolve(tailwindCssPath)).toString();

    if (_tailwindCssContents.length === 0) {
      log.warn('No css found when reading Tailwind configuration css - path', tailwindCssPath);
    }
  } else {
    _tailwindCssContents = _defaultPackageTailwindCss;
  }
}

function setDefaultTailwindConf(tailwindConf?: TailwindConfig) {
  if (tailwindConf) {
    _tailwindConf = tailwindConf;
  } else {
    const twConf = {
      mode: 'jit',
      purge: [],
      darkMode: false,
      plugins: []
    };
    // HACK! Current tailwind types don't have the 'jit' mode field so make the type
    // look as if it is meant to be there
    _tailwindConf = (twConf as unknown) as TailwindConfig;
  }
}

export function setupTailwindConfiguration(tailwindCssPath?: string, tailwindConf?: TailwindConfig) : void {
  setDefaultTailwindCssContents(tailwindCssPath);
  setDefaultTailwindConf(tailwindConf);
}

setDefaultTailwindConf();
setDefaultTailwindCssContents();
