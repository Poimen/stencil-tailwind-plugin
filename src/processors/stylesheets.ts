import ts, { SourceFile, StringLiteral, SyntaxKind } from 'typescript';
import { debug } from '../debug/logger';
import { loadTypescriptCodeFromMemory, makeMatcher, walkTo } from '../helpers/tsFiles';
import { processSourceTextForTailwindInlineClasses, reduceDuplicatedClassesFromFunctionalComponentInjection } from '../helpers/tailwindcss';
import { getAllExternalCssDependencies } from '../store/store';
import { PluginConfigOpts } from '..';

async function transformStyleStatement(opts: PluginConfigOpts, sourceFile: SourceFile, sourceText: string, filename: string) {
  // Stencil produces stylesheet in esm, so need to read and emit back. Stencil outputs the css in the first
  // variable statement in the file. As such there is no clean way of detecting this, so just go grab the
  // first statement
  const stringLiteralPath = [
    makeMatcher(SyntaxKind.SourceFile),
    makeMatcher(SyntaxKind.FirstStatement),
    makeMatcher(SyntaxKind.VariableDeclarationList),
    makeMatcher(SyntaxKind.VariableDeclaration),
    makeMatcher(SyntaxKind.StringLiteral)
  ];

  // Grab any css that needs to be injected by functional components that this component imported
  const injectedCss = getAllExternalCssDependencies(filename).css;
  const stringStyleRewriter = async (cssNode: StringLiteral) => {
    const originalCss = cssNode.text;

    const tailwindClasses = await processSourceTextForTailwindInlineClasses(opts, filename, sourceText, originalCss);
    const reducedClasses = await reduceDuplicatedClassesFromFunctionalComponentInjection(opts, filename, tailwindClasses, injectedCss);

    cssNode.text = reducedClasses;
  };

  const cssNode = walkTo(sourceFile, stringLiteralPath);
  await stringStyleRewriter(cssNode as StringLiteral);

  const printer = ts.createPrinter();
  return printer.printFile(sourceFile);
}

// This function processes css that is contained as a string in a typescript file
// This transform is used when stencil passes a typescript encoded css blob rather than
// the raw file
export function transformCssFromTsxFileFormat(opts: PluginConfigOpts) {
  return async (sourceText: string, filename: string): Promise<string> => {
    debug('[Stylesheets]', 'Processing css from tsx source file:', filename);

    const sourceFile = loadTypescriptCodeFromMemory(sourceText);
    const transformed = await transformStyleStatement(opts, sourceFile, sourceText, filename);

    return transformed;
  };
}

// This function processes a pure css file - i.e. a file with the contents of a css file
// This transform is used during HMR where css files are passed around
export function transformCssFileFormat(opts: PluginConfigOpts) {
  return async (sourceCss: string, filename: string): Promise<string> => {
    debug('[Stylesheets]', 'Processing css source file:', filename);

    const injectedCss = getAllExternalCssDependencies(filename).css;
    const tailwindClasses = await processSourceTextForTailwindInlineClasses(opts, filename, sourceCss, sourceCss);
    const reducedClasses = await reduceDuplicatedClassesFromFunctionalComponentInjection(opts, filename, tailwindClasses, injectedCss);

    return reducedClasses;
  };
}
