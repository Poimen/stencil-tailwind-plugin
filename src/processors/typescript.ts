import path from 'path';
import ts, { ImportDeclaration, BinaryExpression, Identifier, Node, ReturnStatement, SourceFile, StringLiteral, SyntaxKind, ObjectLiteralExpression, PropertyAssignment, VariableDeclaration } from 'typescript';
import { debug, warn, error } from '../debug/logger';
import { loadTypescriptCodeFromMemory, makeMatcher, transformNodeWith, walkAll } from '../helpers/tsFiles';
import { processSourceTextForTailwindInlineClasses } from '../helpers/tailwindcss';
import { registerCssForInjection, registerImportForFile } from '../store/store';

interface TransformationResult {
  text: string;
  transformed: boolean;
}

function reWriteStyleForGetAccessor(sourceFile: SourceFile, css: string) {
  // Stencil creates variable class declarations that use:
  // static get style() { return '...'; }
  // This re-writes the return statement to include tailwind's styles
  const getAccessorPath = [
    makeMatcher(SyntaxKind.SourceFile),
    makeMatcher(SyntaxKind.VariableStatement, { modifier: SyntaxKind.ExportKeyword }),
    makeMatcher(SyntaxKind.VariableDeclarationList),
    makeMatcher(SyntaxKind.VariableDeclaration),
    makeMatcher(SyntaxKind.ClassExpression),
    makeMatcher(SyntaxKind.GetAccessor, { name: 'style', modifier: SyntaxKind.StaticKeyword }),
    makeMatcher(SyntaxKind.Block),
    makeMatcher(SyntaxKind.ReturnStatement)
  ];

  const returnStatementRewriter = (node: Node) => {
    const returnStatement = node as ReturnStatement;
    const originalExpression = returnStatement.expression as Identifier;
    return ts.factory.updateReturnStatement(returnStatement, ts.factory.createIdentifier(`${originalExpression.escapedText} + '${css}'`));
  };

  return transformNodeWith(sourceFile, getAccessorPath, returnStatementRewriter);
}

function reWriteStylePropertyAssignment(sourceFile: SourceFile, css: string) {
  // Stencil creates binary property assignment that uses:
  // ComponentClass.style = '...'
  // This re-writes the assignment to include tailwind's styles
  const binaryExpressionStylePath = [
    makeMatcher(SyntaxKind.SourceFile),
    makeMatcher(SyntaxKind.ExpressionStatement),
    makeMatcher(SyntaxKind.BinaryExpression)
  ];

  const createNewIdentifier = (origValue: ts.__String) => ts.factory.createIdentifier(`${origValue} + '${css}'`);

  const binaryExpressionRewriter = (node: Node) => {
    const binaryStatement = node as BinaryExpression;

    if (ts.isPropertyAccessExpression(binaryStatement.left) && binaryStatement.left.name && binaryStatement.left.name.escapedText === 'style') {
      if (ts.isObjectLiteralExpression(binaryStatement.right)) {
        // Object literals have multiple style properties involved, so add tailwind styles to each one
        const originalExpression = binaryStatement.right as ObjectLiteralExpression;
        const updatedProperties = originalExpression.properties.map(p => {
          const propertyAssignment = p as PropertyAssignment;
          const propertyName = (p.name as Identifier).escapedText as string;
          const originalAssignmentValue = (propertyAssignment.initializer as Identifier).escapedText;
          return ts.factory.createPropertyAssignment(
            propertyName,
            createNewIdentifier(originalAssignmentValue)
          );
        });
        return ts.factory.updateBinaryExpression(
          binaryStatement,
          binaryStatement.left,
          ts.SyntaxKind.EqualsToken,
          ts.factory.createObjectLiteralExpression(updatedProperties)
        );
      } else if (ts.isIdentifier(binaryStatement.right)) {
        const originalExpression = binaryStatement.right as Identifier;
        return ts.factory.updateBinaryExpression(
          binaryStatement,
          binaryStatement.left,
          ts.SyntaxKind.EqualsToken,
          createNewIdentifier(originalExpression.escapedText)
        );
      } else {
        warn('[Typescript]', 'Found a binary expression but', binaryStatement.right.kind, 'is not currently handled');
      }
    }
    error('[Typescript]', 'Found a binary statement, but failed to match it to style!');
    return node;
  };

  return transformNodeWith(sourceFile, binaryExpressionStylePath, binaryExpressionRewriter);
}

function addStylePropertyGetter(sourceFile: SourceFile, css: string) {
  // When there is no style url used, then Stencil does not generate a style property to
  // use for Tailwind style, so add it in
  const getAccessorPath = [
    makeMatcher(SyntaxKind.SourceFile),
    makeMatcher(SyntaxKind.VariableStatement),
    makeMatcher(SyntaxKind.VariableDeclarationList),
    makeMatcher(SyntaxKind.VariableDeclaration, { initializer: SyntaxKind.ClassExpression })
  ];

  let className = '';

  const returnStatementRewriter = (node: Node) => {
    const classDeclaration = node as VariableDeclaration;

    if (classDeclaration.name && classDeclaration.initializer && classDeclaration.initializer.kind === SyntaxKind.ClassExpression) {
      className = (classDeclaration.name as Identifier).escapedText as string;
    }
    return node;
  };

  const result = transformNodeWith(sourceFile, getAccessorPath, returnStatementRewriter);
  if (result.found) {
    const cssAssignment = ts.factory.createIdentifier(`'${css}'`);
    result.fullText = `${result.fullText}${className}.style = ${cssAssignment.escapedText};\n`;
  }

  return result;
}

function registerAllImports(sourceFile: SourceFile, filename: string) {
  const file = path.parse(filename);

  function handleImportDeclaration(node: Node): boolean {
    const importDecl = node as ImportDeclaration;
    const importFilename = importDecl.moduleSpecifier as StringLiteral;
    const importedFile = path.resolve(file.dir, importFilename.text);

    registerImportForFile(filename, importedFile);

    return false;
  }

  walkAll(sourceFile, SyntaxKind.ImportDeclaration, handleImportDeclaration);
}

function transformSourceToIncludeNewTailwindStyles(sourceFile: SourceFile, css: string): TransformationResult {
  let result = reWriteStyleForGetAccessor(sourceFile, css);
  if (!result.found) {
    // If there is no get accessor, fallback to the binary property setter on the class
    result = reWriteStylePropertyAssignment(sourceFile, css);
    if (!result.found) {
      // If no class style, inject it
      result = addStylePropertyGetter(sourceFile, css);
    }
  }

  return {
    text: result.fullText,
    transformed: result.found
  };
}

function preserveTailwindCssEscaping(css: string) {
  // Tailwind JIT syntax needs to escape the '#', '[', ']', etc. in the css output. When writing
  // the escaped characters, the tailwind escaping is lost as the string output doesn't keep the
  // escaping as it doesn't know that they should be escaped in css.
  //
  // This replacement ensures the escaped characters from tailwind are kept escaped so the ending css
  // is correct when applied to styles in css.
  return css.replace(/\\/g, '\\\\');
}

export async function transform(sourceText: string, filename: string): Promise<string> {
  debug('[Typescript]', 'Processing source file:', filename);

  // TODO: remove all import statements so that stencil shadow prop doesn't make classes appear
  const tailwindClasses = await processSourceTextForTailwindInlineClasses(filename);

  if (tailwindClasses.length === 0) {
    // No classes from tailwind to add, just give source back
    return sourceText;
  }

  const escapedCss = preserveTailwindCssEscaping(tailwindClasses);

  const sourceFile = loadTypescriptCodeFromMemory(sourceText);
  registerAllImports(sourceFile, filename);

  const emitResult = transformSourceToIncludeNewTailwindStyles(sourceFile, escapedCss);

  if (!emitResult.transformed) {
    // No placeholder for the css found - register this css for a later style import
    registerCssForInjection(filename, tailwindClasses);
  }

  return emitResult.text;
}
