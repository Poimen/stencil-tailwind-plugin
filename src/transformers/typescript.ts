export async function transform(sourceText: string, fileName: string): Promise<string> {
  fileName = sourceText;
  return fileName;
}
