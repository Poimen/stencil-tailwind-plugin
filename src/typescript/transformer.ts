export async function transform(sourceText: string, fileName: string): Promise<string> {
  // fileName = fileName;
  fileName = sourceText;
  return fileName;
}
