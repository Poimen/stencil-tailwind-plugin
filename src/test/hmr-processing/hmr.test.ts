import { PluginCtx, PluginTransformResults } from '@stencil/core/internal';
import { configurePluginOptions, PluginConfigDefaults } from '../../config/pluginConfiguration';
import { tailwindHMR } from '../../index';
import { transform } from '../../processors/typescript';
import { loadTestComponent } from '../utils';

describe('hmr', () => {
  it('given style css should process @apply rules', async () => {
    // Arrange
    let convertedCss: string | undefined;
    const style = loadTestComponent('hmr-processing', 'style.css');
    const filePath = style.path.replace(/\\/g, '/');

    const plugin = tailwindHMR(PluginConfigDefaults.DEFAULT);
    // Act
    if (plugin.transform) {
      const transform = (await plugin.transform(style.text, filePath, { } as PluginCtx)) as PluginTransformResults;
      convertedCss = transform.code;
    }
    // Assert
    expect(convertedCss).toMatchSnapshot();
  });

  it('given style css with dependents should process @apply rules with dependents css attached', async () => {
    // Arrange
    let convertedCss: string | undefined;
    let deps: string[];
    const styleForHMR = loadTestComponent('hmr-processing', 'style.css');
    const depLoadedFile = loadTestComponent('hmr-processing', 'css-hmr-component-tailwind.tsx');

    await transform(configurePluginOptions(PluginConfigDefaults.DEFAULT))(depLoadedFile.text, depLoadedFile.path);

    const filePath = styleForHMR.path.replace(/\\/g, '/');

    const plugin = tailwindHMR(PluginConfigDefaults.DEFAULT);

    // Act
    if (plugin.transform) {
      const transform = (await plugin.transform(styleForHMR.text, filePath, { } as PluginCtx)) as PluginTransformResults;
      convertedCss = transform.code;
      deps = transform.dependencies ?? [];

      // Assert
      expect(convertedCss).toMatchSnapshot();
      expect(deps.length).toBe(1);
      expect(deps[0]).toContain('css-hmr-component-tailwind.tsx');
    } else {
      fail('should have a transform function');
    }
  });
});
