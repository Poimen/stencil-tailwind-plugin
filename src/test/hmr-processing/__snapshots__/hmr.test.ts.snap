// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hmr given style css should process @apply rules 1`] = `"/*! tailwindcss v4.1.1 | MIT License | https://tailwindcss.com */@layer theme{:host,:root{--font-sans:ui-sans-serif,system-ui,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";--font-mono:ui-monospace,SFMono-Regular,Menlo,Monaco,Consolas,"Liberation Mono","Courier New",monospace;--default-transition-duration:.15s;--default-transition-timing-function:cubic-bezier(.4,0,.2,1);--default-font-family:var(--font-sans);--default-mono-font-family:var(--font-mono)}}@layer base{*,::backdrop,:after,:before{border:0 solid;box-sizing:border-box;margin:0;padding:0;--tw-rotate-x:rotateX(0);--tw-rotate-y:rotateY(0);--tw-rotate-z:rotateZ(0);--tw-skew-x:skewX(0);--tw-skew-y:skewY(0);--tw-duration:initial;--tw-scale-x:1;--tw-scale-y:1;--tw-scale-z:1}::file-selector-button{appearance:button;background-color:#0000;border:0 solid;border-radius:0;box-sizing:border-box;color:inherit;font:inherit;font-feature-settings:inherit;font-variation-settings:inherit;letter-spacing:inherit;margin:0;margin-inline-end:4px;opacity:1;padding:0}:host,html{-webkit-text-size-adjust:100%;font-family:var(--default-font-family,ui-sans-serif,system-ui,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji");font-feature-settings:var(--default-font-feature-settings,normal);font-variation-settings:var(--default-font-variation-settings,normal);line-height:1.5;tab-size:4;-webkit-tap-highlight-color:transparent}hr{border-top-width:1px;color:inherit;height:0}abbr:where([title]){-webkit-text-decoration:underline dotted;text-decoration:underline dotted}h1,h2,h3,h4,h5,h6{font-size:inherit;font-weight:inherit}a{color:inherit;-webkit-text-decoration:inherit;text-decoration:inherit}b,strong{font-weight:bolder}code,kbd,pre,samp{font-family:var(--default-mono-font-family,ui-monospace,SFMono-Regular,Menlo,Monaco,Consolas,"Liberation Mono","Courier New",monospace);font-feature-settings:var(--default-mono-font-feature-settings,normal);font-size:1em;font-variation-settings:var(--default-mono-font-variation-settings,normal)}small{font-size:80%}sub,sup{font-size:75%;line-height:0;position:relative;vertical-align:baseline}sub{bottom:-.25em}sup{top:-.5em}table{border-collapse:collapse;border-color:inherit;text-indent:0}:-moz-focusring{outline:auto}progress{vertical-align:baseline}summary{display:list-item}menu,ol,ul{list-style:none}audio,canvas,embed,iframe,img,object,svg,video{display:block;vertical-align:middle}img,video{height:auto;max-width:100%}button,input,optgroup,select,textarea{background-color:#0000;border-radius:0;color:inherit;font:inherit;font-feature-settings:inherit;font-variation-settings:inherit;letter-spacing:inherit;opacity:1}:where(select:is([multiple],[size])) optgroup{font-weight:bolder}:where(select:is([multiple],[size])) optgroup option{padding-inline-start:20px}::placeholder{opacity:1}@supports (not ((-webkit-appearance:-apple-pay-button))) or (contain-intrinsic-size:1px){::placeholder{color:color-mix(in oklab,currentColor 50%,transparent)}}textarea{resize:vertical}::-webkit-search-decoration{-webkit-appearance:none}::-webkit-date-and-time-value{min-height:1lh;text-align:inherit}::-webkit-datetime-edit{display:inline-flex;padding-block:0}::-webkit-datetime-edit-fields-wrapper{padding:0}::-webkit-datetime-edit-month-field,::-webkit-datetime-edit-year-field{padding-block:0}::-webkit-datetime-edit-day-field,::-webkit-datetime-edit-hour-field{padding-block:0}::-webkit-datetime-edit-minute-field,::-webkit-datetime-edit-second-field{padding-block:0}::-webkit-datetime-edit-meridiem-field,::-webkit-datetime-edit-millisecond-field{padding-block:0}:-moz-ui-invalid{box-shadow:none}button,input:where([type=button],[type=reset],[type=submit]){appearance:button}::-webkit-inner-spin-button,::-webkit-outer-spin-button{height:auto}[hidden]:where(:not([hidden=until-found])){display:none!important}}@layer components;@layer utilities{.flex{display:flex}.w-6\\/12{width:50%}.transform{transform:var(--tw-rotate-x)var(--tw-rotate-y)var(--tw-rotate-z)var(--tw-skew-x)var(--tw-skew-y)}.flex-col{flex-direction:column}.bg-\\[\\#343da3\\]{background-color:#343da3}.text-\\[3em\\]{font-size:3em}.text-\\[\\#dedede\\]{color:#dedede}.transition{transition-duration:var(--tw-duration,var(--default-transition-duration));transition-property:color,background-color,border-color,outline-color,text-decoration-color,fill,stroke,--tw-gradient-from,--tw-gradient-via,--tw-gradient-to,opacity,box-shadow,transform,translate,scale,rotate,filter,-webkit-backdrop-filter,backdrop-filter;transition-timing-function:var(--tw-ease,var(--default-transition-timing-function))}.duration-500{--tw-duration:.5s;transition-duration:.5s}}:host{display:block;text-overflow:clip}@property --tw-rotate-x{syntax:"*";inherits:false;initial-value:rotateX(0)}@property --tw-rotate-y{syntax:"*";inherits:false;initial-value:rotateY(0)}@property --tw-rotate-z{syntax:"*";inherits:false;initial-value:rotateZ(0)}@property --tw-skew-x{syntax:"*";inherits:false;initial-value:skewX(0)}@property --tw-skew-y{syntax:"*";inherits:false;initial-value:skewY(0)}@property --tw-duration{syntax:"*";inherits:false}@property --tw-scale-x{syntax:"*";inherits:false;initial-value:1}@property --tw-scale-y{syntax:"*";inherits:false;initial-value:1}@property --tw-scale-z{syntax:"*";inherits:false;initial-value:1}@media (hover:hover){.hover\\:scale-110:hover{--tw-scale-x:110%;--tw-scale-y:110%;--tw-scale-z:110%;scale:var(--tw-scale-x)var(--tw-scale-y)}}"`;

exports[`hmr given style css with dependents should process @apply rules with dependents css attached 1`] = `"/*! tailwindcss v4.1.1 | MIT License | https://tailwindcss.com */@layer theme{:host,:root{--font-sans:ui-sans-serif,system-ui,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";--font-mono:ui-monospace,SFMono-Regular,Menlo,Monaco,Consolas,"Liberation Mono","Courier New",monospace;--default-transition-duration:.15s;--default-transition-timing-function:cubic-bezier(.4,0,.2,1);--default-font-family:var(--font-sans);--default-mono-font-family:var(--font-mono)}}@layer base{*,::backdrop,:after,:before{border:0 solid;box-sizing:border-box;margin:0;padding:0;--tw-rotate-x:rotateX(0);--tw-rotate-y:rotateY(0);--tw-rotate-z:rotateZ(0);--tw-skew-x:skewX(0);--tw-skew-y:skewY(0);--tw-duration:initial;--tw-scale-x:1;--tw-scale-y:1;--tw-scale-z:1}::file-selector-button{appearance:button;background-color:#0000;border:0 solid;border-radius:0;box-sizing:border-box;color:inherit;font:inherit;font-feature-settings:inherit;font-variation-settings:inherit;letter-spacing:inherit;margin:0;margin-inline-end:4px;opacity:1;padding:0}:host,html{-webkit-text-size-adjust:100%;font-family:var(--default-font-family,ui-sans-serif,system-ui,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji");font-feature-settings:var(--default-font-feature-settings,normal);font-variation-settings:var(--default-font-variation-settings,normal);line-height:1.5;tab-size:4;-webkit-tap-highlight-color:transparent}hr{border-top-width:1px;color:inherit;height:0}abbr:where([title]){-webkit-text-decoration:underline dotted;text-decoration:underline dotted}h1,h2,h3,h4,h5,h6{font-size:inherit;font-weight:inherit}a{color:inherit;-webkit-text-decoration:inherit;text-decoration:inherit}b,strong{font-weight:bolder}code,kbd,pre,samp{font-family:var(--default-mono-font-family,ui-monospace,SFMono-Regular,Menlo,Monaco,Consolas,"Liberation Mono","Courier New",monospace);font-feature-settings:var(--default-mono-font-feature-settings,normal);font-size:1em;font-variation-settings:var(--default-mono-font-variation-settings,normal)}small{font-size:80%}sub,sup{font-size:75%;line-height:0;position:relative;vertical-align:baseline}sub{bottom:-.25em}sup{top:-.5em}table{border-collapse:collapse;border-color:inherit;text-indent:0}:-moz-focusring{outline:auto}progress{vertical-align:baseline}summary{display:list-item}menu,ol,ul{list-style:none}audio,canvas,embed,iframe,img,object,svg,video{display:block;vertical-align:middle}img,video{height:auto;max-width:100%}button,input,optgroup,select,textarea{background-color:#0000;border-radius:0;color:inherit;font:inherit;font-feature-settings:inherit;font-variation-settings:inherit;letter-spacing:inherit;opacity:1}:where(select:is([multiple],[size])) optgroup{font-weight:bolder}:where(select:is([multiple],[size])) optgroup option{padding-inline-start:20px}::placeholder{opacity:1}@supports (not ((-webkit-appearance:-apple-pay-button))) or (contain-intrinsic-size:1px){::placeholder{color:color-mix(in oklab,currentColor 50%,transparent)}}textarea{resize:vertical}::-webkit-search-decoration{-webkit-appearance:none}::-webkit-date-and-time-value{min-height:1lh;text-align:inherit}::-webkit-datetime-edit{display:inline-flex;padding-block:0}::-webkit-datetime-edit-fields-wrapper{padding:0}::-webkit-datetime-edit-month-field,::-webkit-datetime-edit-year-field{padding-block:0}::-webkit-datetime-edit-day-field,::-webkit-datetime-edit-hour-field{padding-block:0}::-webkit-datetime-edit-minute-field,::-webkit-datetime-edit-second-field{padding-block:0}::-webkit-datetime-edit-meridiem-field,::-webkit-datetime-edit-millisecond-field{padding-block:0}:-moz-ui-invalid{box-shadow:none}button,input:where([type=button],[type=reset],[type=submit]){appearance:button}::-webkit-inner-spin-button,::-webkit-outer-spin-button{height:auto}[hidden]:where(:not([hidden=until-found])){display:none!important}}@layer components;@layer utilities{.flex{display:flex}.w-6\\/12{width:50%}.transform{transform:var(--tw-rotate-x)var(--tw-rotate-y)var(--tw-rotate-z)var(--tw-skew-x)var(--tw-skew-y)}.flex-col{flex-direction:column}.bg-\\[\\#343da3\\]{background-color:#343da3}.text-\\[3em\\]{font-size:3em}.text-\\[\\#dedede\\]{color:#dedede}.transition{transition-duration:var(--tw-duration,var(--default-transition-duration));transition-property:color,background-color,border-color,outline-color,text-decoration-color,fill,stroke,--tw-gradient-from,--tw-gradient-via,--tw-gradient-to,opacity,box-shadow,transform,translate,scale,rotate,filter,-webkit-backdrop-filter,backdrop-filter;transition-timing-function:var(--tw-ease,var(--default-transition-timing-function))}.duration-500{--tw-duration:.5s;transition-duration:.5s}} /*! tailwindcss v4.1.1 | MIT License | https://tailwindcss.com */@layer components;:host{display:block;text-overflow:clip}@property --tw-rotate-x{syntax:"*";inherits:false;initial-value:rotateX(0)}@property --tw-rotate-y{syntax:"*";inherits:false;initial-value:rotateY(0)}@property --tw-rotate-z{syntax:"*";inherits:false;initial-value:rotateZ(0)}@property --tw-skew-x{syntax:"*";inherits:false;initial-value:skewX(0)}@property --tw-skew-y{syntax:"*";inherits:false;initial-value:skewY(0)}@property --tw-duration{syntax:"*";inherits:false}@property --tw-scale-x{syntax:"*";inherits:false;initial-value:1}@property --tw-scale-y{syntax:"*";inherits:false;initial-value:1}@property --tw-scale-z{syntax:"*";inherits:false;initial-value:1}@media (hover:hover){.hover\\:scale-110:hover{--tw-scale-x:110%;--tw-scale-y:110%;--tw-scale-z:110%;scale:var(--tw-scale-x)var(--tw-scale-y)}}"`;
